archive/build_artifacts2/staging/share/man/man1/tset.1
.\"***************************************************************************
.\" Copyright 2018-2021,2022 Thomas E. Dickey                                *
.\" Copyright 1998-2016,2017 Free Software Foundation, Inc.                  *
.\"                                                                          *
.\" Permission is hereby granted, free of charge, to any person obtaining a  *
.\" copy of this software and associated documentation files (the            *
.\" "Software"), to deal in the Software without restriction, including      *
.\" without limitation the rights to use, copy, modify, merge, publish,      *
.\" distribute, distribute with modifications, sublicense, and/or sell       *
.\" copies of the Software, and to permit persons to whom the Software is    *
.\" furnished to do so, subject to the following conditions:                 *
.\"                                                                          *
.\" The above copyright notice and this permission notice shall be included  *
.\" in all copies or substantial portions of the Software.                   *
.\"                                                                          *
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS  *
.\" OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF               *
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.   *
.\" IN NO EVENT SHALL THE ABOVE COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,   *
.\" DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR    *
.\" OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR    *
.\" THE USE OR OTHER DEALINGS IN THE SOFTWARE.                               *
.\"                                                                          *
.\" Except as contained in this notice, the name(s) of the above copyright   *
.\" holders shall not be used in advertising or otherwise to promote the     *
.\" sale, use or other dealings in this Software without prior written       *
.\" authorization.                                                           *
.\"***************************************************************************
.\"
.\" $Id: tset.1,v 1.62 2022/02/12 20:02:20 tom Exp $
.TH tset 1 ""
.ie \n(.g .ds `` \(lq
.el       .ds `` ``
.ie \n(.g .ds '' \(rq
.el       .ds '' ''
.de bP
.ie n  .IP \(bu 4
.el    .IP \(bu 2
..
.SH NAME
\fBtset\fP, \fBreset\fP \- terminal initialization
.SH SYNOPSIS
\fBtset\fP [\fB\-IQVcqrsw\fP] [\fB\-\fP] [\fB\-e\fP \fIch\fP] [\fB\-i\fP \fIch\fP] [\fB\-k\fP \fIch\fP] [\fB\-m\fP \fImapping\fP] [\fIterminal\fP]
.br
\fBreset\fP [\fB\-IQVcqrsw\fP] [\fB\-\fP] [\fB\-e\fP \fIch\fP] [\fB\-i\fP \fIch\fP] [\fB\-k\fP \fIch\fP] [\fB\-m\fP \fImapping\fP] [\fIterminal\fP]
.SH DESCRIPTION
.SS tset - initialization
This program initializes terminals.
.PP
First, \fBtset\fP retrieves the current terminal mode settings
for your terminal.
It does this by successively testing
.bP
the standard error,
.bP
standard output,
.bP
standard input and
.bP
ultimately \*(``/dev/tty\*(''
.PP
to obtain terminal settings.
Having retrieved these settings, \fBtset\fP remembers which
file descriptor to use when updating settings.
.PP
Next, \fBtset\fP determines the type of terminal that you are using.
This determination is done as follows, using the first terminal type found.
.PP
1. The \fBterminal\fP argument specified on the command line.
.PP
2. The value of the \fBTERM\fP environmental variable.
.PP
3. (BSD systems only.) The terminal type associated with the standard
error output device in the \fI/etc/ttys\fP file.
(On System\-V-like UNIXes and systems using that convention,
\fBgetty\fP(8) does this job by setting
\fBTERM\fP according to the type passed to it by \fI/etc/inittab\fP.)
.PP
4. The default terminal type, \*(``unknown\*(''.
.PP
If the terminal type was not specified on the command-line, the \fB\-m\fP
option mappings are then applied (see the section
.B TERMINAL TYPE MAPPING
for more information).
Then, if the terminal type begins with a question mark (\*(``?\*(''), the
user is prompted for confirmation of the terminal type.
An empty
response confirms the type, or, another type can be entered to specify
a new type.
Once the terminal type has been determined,
the terminal description for the terminal is retrieved.
If no terminal description is found
for the type, the user is prompted for another terminal type.
.PP
Once the terminal description is retrieved,
.bP
if the \*(``\fB\-w\fP\*('' option is enabled, \fBtset\fP may update
the terminal's window size.
.IP
If the window size cannot be obtained from the operating system,
but the terminal description (or environment, e.g., \fBLINES\fP
and \fBCOLUMNS\fP variables specify this),
use this to set the operating system's notion of the window size.
.bP
if the \*(``\fB\-c\fP\*('' option is enabled,
the backspace, interrupt and line kill characters
(among many other things) are set
.bP
unless the \*(``\fB\-I\fP\*('' option is enabled,
the terminal
and tab \fIinitialization\fP strings are sent to the standard error output,
and \fBtset\fP waits one second (in case a hardware reset was issued).
.bP
Finally, if the erase, interrupt and line kill characters have changed,
or are not set to their default values, their values are displayed to the
standard error output.
.SS reset - reinitialization
.PP
When invoked as \fBreset\fP, \fBtset\fP sets the terminal
modes to \*(``sane\*('' values:
.bP
sets cooked and echo modes,
.bP
turns off cbreak and raw modes,
.bP
turns on newline translation and
.bP
resets any unset special characters to their default values
.PP
before
doing the terminal initialization described above.
Also, rather than using the terminal \fIinitialization\fP strings,
it uses the terminal \fIreset\fP strings.
.PP
The \fBreset\fP command is useful
after a program dies leaving a terminal in an abnormal state:
.bP
you may have to type
.sp
    \fI<LF>\fBreset\fI<LF>\fR
.sp
(the line-feed character is normally control-J) to get the terminal
to work, as carriage-return may no longer work in the abnormal state.
.bP
Also, the terminal will often not echo the command.
.SH OPTIONS
.PP
The options are as follows:
.TP 5
.B \-c
Set control characters and modes.
.TP 5
.BI \-e\  ch
Set the erase character to \fIch\fP.
.TP
.B \-I
Do not send the terminal or tab initialization strings to the terminal.
.TP
.BI \-i\  ch
Set the interrupt character to \fIch\fP.
.TP
.BI \-k\  ch
Set the line kill character to \fIch\fP.
.TP
.BI \-m\  mapping
Specify a mapping from a port type to a terminal.
See the section
.B TERMINAL TYPE MAPPING
for more information.
.TP
.B \-Q
Do not display any values for the erase, interrupt and line kill characters.
Normally \fBtset\fP displays the values for control characters which
differ from the system's default values.
.TP
.B \-q
The terminal type is displayed to the standard output, and the terminal is
not initialized in any way.
The option \*(``\-\*('' by itself is equivalent but archaic.
.TP
.B \-r
Print the terminal type to the standard error output.
.TP
.B \-s
Print the sequence of shell commands to initialize the environment variable
\fBTERM\fP to the standard output.
See the section
.B SETTING THE ENVIRONMENT
for details.
.TP
.B \-V
reports the version of ncurses which was used in this program, and exits.
.TP
.B \-w
Resize the window to match the size deduced via \fBsetupterm\fP(3X).
Normally this has no effect,
unless \fBsetupterm\fP is not able to detect the window size.
.PP
The arguments for the \fB\-e\fP, \fB\-i\fP, and \fB\-k\fP
options may either be entered as actual characters
or by using the \*(``hat\*(''
notation, i.e., control-h may be specified as \*(``^H\*('' or \*(``^h\*(''.
.PP
If neither \fB\-c\fP or \fB\-w\fP is given, both options are assumed.
.
.SH SETTING THE ENVIRONMENT
It is often desirable to enter the terminal type and information about
the terminal's capabilities into the shell's environment.
This is done using the \fB\-s\fP option.
.PP
When the \fB\-s\fP option is specified, the commands to enter the information
into the shell's environment are written to the standard output.
If
the \fBSHELL\fP environmental variable ends in \*(``csh\*('', the commands
are for \fBcsh\fP, otherwise, they are for \fBsh\fP(1).
Note, the \fBcsh\fP commands set and unset the shell variable
\fBnoglob\fP, leaving it unset.
The following line in the \fB.login\fP
or \fB.profile\fP files will initialize the environment correctly:
.sp
    eval \`tset \-s options ... \`
.
.SH TERMINAL TYPE MAPPING
When the terminal is not hardwired into the system (or the current
system information is incorrect) the terminal type derived from the
\fI/etc/ttys\fP file or the \fBTERM\fP environmental variable is often
something generic like \fBnetwork\fP, \fBdialup\fP, or \fBunknown\fP.
When \fBtset\fP is used in a startup script it is often desirable to
provide information about the type of terminal used on such ports.
.PP
The \fB\-m\fP options maps
from some set of conditions to a terminal type, that is, to
tell \fBtset\fP
\*(``If I'm on this port at a particular speed,
guess that I'm on that kind of terminal\*(''.
.PP
The argument to the \fB\-m\fP option consists of an optional port type, an
optional operator, an optional baud rate specification, an optional
colon (\*(``:\*('') character and a terminal type.
The port type is a
string (delimited by either the operator or the colon character).
The operator may be any combination of
\*(``>\*('',
\*(``<\*('',
\*(``@\*('',
and \*(``!\*('';
\*(``>\*('' means greater than,
\*(``<\*('' means less than,
\*(``@\*('' means equal to and
\*(``!\*('' inverts the sense of the test.
The baud rate is specified as a number and is compared with the speed
of the standard error output (which should be the control terminal).
The terminal type is a string.
.PP
If the terminal type is not specified on the command line, the \fB\-m\fP
mappings are applied to the terminal type.
If the port type and baud
rate match the mapping, the terminal type specified in the mapping
replaces the current type.
If more than one mapping is specified, the
first applicable mapping is used.
.PP
For example, consider the following mapping: \fBdialup>9600:vt100\fP.
The port type is dialup , the operator is >, the baud rate
specification is 9600, and the terminal type is vt100.
The result of
this mapping is to specify that if the terminal type is \fBdialup\fP,
and the baud rate is greater than 9600 baud, a terminal type of
\fBvt100\fP will be used.
.PP
If no baud rate is specified, the terminal type will match any baud rate.
If no port type is specified, the terminal type will match any port type.
For example, \fB\-m dialup:vt100 \-m :?xterm\fP
will cause any dialup port, regardless of baud rate, to match the terminal
type vt100, and any non-dialup port type to match the terminal type ?xterm.
Note, because of the leading question mark, the user will be
queried on a default port as to whether they are actually using an xterm
terminal.
.PP
No whitespace characters are permitted in the \fB\-m\fP option argument.
Also, to avoid problems with meta-characters, it is suggested that the
entire \fB\-m\fP option argument be placed within single quote characters,
and that \fBcsh\fP users insert a backslash character (\*(``\e\*('') before
any exclamation marks (\*(``!\*('').
.SH HISTORY
.PP
A \fBreset\fP command appeared in 1BSD (March 1978), written by Kurt Shoens.
This program set the \fIerase\fP and \fIkill\fP characters
to \fB^H\fP (backspace) and \fB@\fP respectively.
Mark Horton improved that in 3BSD (October 1979), adding
\fIintr\fP, \fIquit\fP, \fIstart\fP/\fIstop\fP and \fIeof\fP characters
as well as changing the program to avoid modifying any user settings.
That version of \fBreset\fP did not use the termcap database.
.PP
A separate \fBtset\fP command was provided in 1BSD by Eric Allman,
using the termcap database.
Allman's comments in the source code indicate
that he began work in October 1977,
continuing development over the next few years.
.PP
According to comments in the source code,
the \fBtset\fP program was modified in September 1980,
to use logic copied from the 3BSD \*(``reset\*(''
when it was invoked as \fBreset\fP.
This version appeared in 4.1cBSD, late in 1982.
.PP
Other developers (e.g., Keith Bostic and Jim Bloom)
continued to modify \fBtset\fP until 4.4BSD was released in 1993.
.PP
The \fBncurses\fP implementation
was lightly adapted from the 4.4BSD sources for a terminfo environment by Eric
S. Raymond <esr@snark.thyrsus.com>.
.SH COMPATIBILITY
.PP
Neither IEEE Std 1003.1/The Open Group Base Specifications Issue 7
(POSIX.1-2008) nor
X/Open Curses Issue 7 documents \fBtset\fP or \fBreset\fP.
.PP
The AT&T \fBtput\fP utility (AIX, HPUX, Solaris)
incorporated the terminal-mode manipulation as well as termcap-based features
such as resetting tabstops from \fBtset\fP in BSD (4.1c),
presumably with the intention of making \fBtset\fP obsolete.
However, each of those systems still provides \fBtset\fP.
In fact, the commonly-used \fBreset\fP utility
is always an alias for \fBtset\fP.
.PP
The \fBtset\fP utility provides for backward-compatibility with BSD
environments (under most modern UNIXes, \fB/etc/inittab\fP and \fBgetty\fP(8)
can set \fBTERM\fP appropriately for each dial-up line; this obviates what was
\fBtset\fP's most important use).
This implementation behaves like 4.4BSD
\fBtset\fP, with a few exceptions specified here.
.PP
A few options are different
because the \fBTERMCAP\fP variable
is no longer supported under terminfo-based \fBncurses\fP:
.bP
The \fB\-S\fP option of BSD \fBtset\fP no longer works;
it prints an error message to the standard error and dies.
.bP
The \fB\-s\fP option only sets \fBTERM\fP, not \fBTERMCAP\fP.
.PP
There was an undocumented 4.4BSD feature
that invoking \fBtset\fP via a link named
\*(``TSET\*('' (or via any other name beginning with an upper-case letter)
set the terminal to use upper-case only.
This feature has been omitted.
.PP
The \fB\-A\fP, \fB\-E\fP, \fB\-h\fP, \fB\-u\fP and \fB\-v\fP
options were deleted from the \fBtset\fP
utility in 4.4BSD.
None of them were documented in 4.3BSD and all are
of limited utility at best.
The \fB\-a\fP, \fB\-d\fP, and \fB\-p\fP options are similarly
not documented or useful, but were retained as they appear to be in
widespread use.
It is strongly recommended that any usage of these
three options be changed to use the \fB\-m\fP option instead.
The \fB\-a\fP, \fB\-d\fP, and \fB\-p\fP options
are therefore omitted from the usage summary above.
.PP
Very old systems, e.g., 3BSD, used a different terminal driver which
was replaced in 4BSD in the early 1980s.
To accommodate these older systems, the 4BSD \fBtset\fP provided a
\fB\-n\fP option to specify that the new terminal driver should be used.
This implementation does not provide that choice.
.PP
It is still permissible to specify the \fB\-e\fP, \fB\-i\fP,
and \fB\-k\fP options without arguments,
although it is strongly recommended that such usage be fixed to
explicitly specify the character.
.PP
As of 4.4BSD,
executing \fBtset\fP as \fBreset\fP no longer implies the \fB\-Q\fP option.
Also, the interaction between the \- option and the \fIterminal\fP
argument in some historic implementations of \fBtset\fP has been removed.
.PP
The \fB\-c\fP and \fB\-w\fP options are not found in earlier implementations.
However, a different window size-change feature was provided in 4.4BSD.
.bP
In 4.4BSD, \fBtset\fP uses the window size from the termcap description
to set the window size if \fBtset\fP is not able to obtain the window
size from the operating system.
.bP
In ncurses, \fBtset\fP obtains the window size using
\fBsetupterm\fP, which may be from
the operating system,
the \fBLINES\fP and \fBCOLUMNS\fP environment variables or
the terminal description.
.PP
Obtaining the window size from the terminal description is common to
both implementations, but considered obsolescent.
Its only practical use is for hardware terminals.
Generally speaking, a window size would be unset only if there were
some problem obtaining the value from the operating system
(and \fBsetupterm\fP would still fail).
For that reason, the \fBLINES\fP and \fBCOLUMNS\fP environment variables
may be useful for working around window-size problems.
Those have the drawback that if the window is resized,
those variables must be recomputed and reassigned.
To do this more easily, use the \fBresize\fP(1) program.
.SH ENVIRONMENT
The \fBtset\fP command uses these environment variables:
.TP 5
SHELL
tells \fBtset\fP whether to initialize \fBTERM\fP using \fBsh\fP(1) or
\fBcsh\fP(1) syntax.
.TP 5
TERM
Denotes your terminal type.
Each terminal type is distinct, though many are similar.
.TP 5
TERMCAP
may denote the location of a termcap database.
If it is not an absolute pathname, e.g., begins with a \*(``/\*('',
\fBtset\fP removes the variable from the environment before looking
for the terminal description.
.SH FILES
.TP 5
/etc/ttys
system port name to terminal type mapping database (BSD versions only).
.TP
//share/terminfo
terminal capability database
.SH SEE ALSO
.hy 0
\fBcsh\fP(1),
\fBsh\fP(1),
\fBstty\fP(1),
\fBterminfo\fP(3NCURSES),
\fBtty\fP(4),
\fBterminfo\fP(5),
\fBttys\fP(4),
\fBenviron\fP(7)
.hy
.PP
This describes \fBncurses\fP
version 6.4 (patch 20221231).


oss_sovereignty/ncurses-6.4/progs/tset.c
 

 

 

 
 

#include <reset_cmd.h>
#include <termcap.h>
#include <transform.h>
#include <tty_settings.h>

#if HAVE_GETTTYNAM
#include <ttyent.h>
#endif
#ifdef NeXT
char *ttyname(int fd);
#endif

MODULE_ID("$Id: tset.c,v 1.131 2021/12/04 23:02:13 tom Exp $")

#ifndef environ
extern char **environ;
#endif

const char *_nc_progname = "tset";

#define LOWERCASE(c) ((isalpha(UChar(c)) && isupper(UChar(c))) ? tolower(UChar(c)) : (c))

static GCC_NORETURN void exit_error(void);

static int
CaselessCmp(const char *a, const char *b)
{				 
    while (*a && *b) {
	int cmp = LOWERCASE(*a) - LOWERCASE(*b);
	if (cmp != 0)
	    break;
	a++, b++;
    }
    return LOWERCASE(*a) - LOWERCASE(*b);
}

static GCC_NORETURN void
exit_error(void)
{
    restore_tty_settings();
    (void) fprintf(stderr, "\n");
    fflush(stderr);
    ExitProgram(EXIT_FAILURE);
     
}

static GCC_NORETURN void
err(const char *fmt, ...)
{
    va_list ap;
    va_start(ap, fmt);
    (void) fprintf(stderr, "%s: ", _nc_progname);
    (void) vfprintf(stderr, fmt, ap);
    va_end(ap);
    exit_error();
     
}

static GCC_NORETURN void
failed(const char *msg)
{
    char temp[BUFSIZ];
    size_t len = strlen(_nc_progname) + 2;

    if ((int) len < (int) sizeof(temp) - 12) {
	_nc_STRCPY(temp, _nc_progname, sizeof(temp));
	_nc_STRCAT(temp, ": ", sizeof(temp));
    } else {
	_nc_STRCPY(temp, "tset: ", sizeof(temp));
    }
    _nc_STRNCAT(temp, msg, sizeof(temp), sizeof(temp) - strlen(temp) - 2);
    perror(temp);
    exit_error();
     
}

 
static const char *
askuser(const char *dflt)
{
    static char answer[256];

     
    clearerr(stdin);
    if (feof(stdin) || ferror(stdin)) {
	(void) fprintf(stderr, "\n");
	exit_error();
	 
    }

    for (;;) {
	char *p;

	if (dflt)
	    (void) fprintf(stderr, "Terminal type? [%s] ", dflt);
	else
	    (void) fprintf(stderr, "Terminal type? ");
	(void) fflush(stderr);

	if (fgets(answer, sizeof(answer), stdin) == 0) {
	    if (dflt == 0) {
		exit_error();
		 
	    }
	    return (dflt);
	}

	if ((p = strchr(answer, '\n')) != 0)
	    *p = '\0';
	if (answer[0])
	    return (answer);
	if (dflt != 0)
	    return (dflt);
    }
}

 

 
#define	GT		0x01
#define	EQ		0x02
#define	LT		0x04
#define	NOT		0x08
#define	GE		(GT | EQ)
#define	LE		(LT | EQ)

typedef struct map {
    struct map *next;		 
    const char *porttype;	 
    const char *type;		 
    int conditional;		 
    int speed;			 
} MAP;

static MAP *cur, *maplist;

#define DATA(name,value) { { name }, value }

typedef struct speeds {
    const char string[8];
    int speed;
} SPEEDS;

#if defined(EXP_WIN32_DRIVER)
static const SPEEDS speeds[] =
{
    {"0", 0}
};
#else
static const SPEEDS speeds[] =
{
    DATA("0", B0),
    DATA("50", B50),
    DATA("75", B75),
    DATA("110", B110),
    DATA("134", B134),
    DATA("134.5", B134),
    DATA("150", B150),
    DATA("200", B200),
    DATA("300", B300),
    DATA("600", B600),
    DATA("1200", B1200),
    DATA("1800", B1800),
    DATA("2400", B2400),
    DATA("4800", B4800),
    DATA("9600", B9600),
     
#ifdef B19200
    DATA("19200", B19200),
#endif
#ifdef B38400
    DATA("38400", B38400),
#endif
#ifdef B19200
    DATA("19200", B19200),
#endif
#ifdef B38400
    DATA("38400", B38400),
#endif
#ifdef B19200
    DATA("19200", B19200),
#else
#ifdef EXTA
    DATA("19200", EXTA),
#endif
#endif
#ifdef B38400
    DATA("38400", B38400),
#else
#ifdef EXTB
    DATA("38400", EXTB),
#endif
#endif
#ifdef B57600
    DATA("57600", B57600),
#endif
#ifdef B76800
    DATA("76800", B57600),
#endif
#ifdef B115200
    DATA("115200", B115200),
#endif
#ifdef B153600
    DATA("153600", B153600),
#endif
#ifdef B230400
    DATA("230400", B230400),
#endif
#ifdef B307200
    DATA("307200", B307200),
#endif
#ifdef B460800
    DATA("460800", B460800),
#endif
#ifdef B500000
    DATA("500000", B500000),
#endif
#ifdef B576000
    DATA("576000", B576000),
#endif
#ifdef B921600
    DATA("921600", B921600),
#endif
#ifdef B1000000
    DATA("1000000", B1000000),
#endif
#ifdef B1152000
    DATA("1152000", B1152000),
#endif
#ifdef B1500000
    DATA("1500000", B1500000),
#endif
#ifdef B2000000
    DATA("2000000", B2000000),
#endif
#ifdef B2500000
    DATA("2500000", B2500000),
#endif
#ifdef B3000000
    DATA("3000000", B3000000),
#endif
#ifdef B3500000
    DATA("3500000", B3500000),
#endif
#ifdef B4000000
    DATA("4000000", B4000000),
#endif
};
#undef DATA
#endif

static int
tbaudrate(char *rate)
{
    const SPEEDS *sp = 0;
    size_t n;

     
    if (*rate == 'B')
	++rate;

    for (n = 0; n < SIZEOF(speeds); ++n) {
	if (n > 0 && (speeds[n].speed <= speeds[n - 1].speed)) {
	     
	    break;
	}
	if (!CaselessCmp(rate, speeds[n].string)) {
	    sp = speeds + n;
	    break;
	}
    }
    if (sp == 0)
	err("unknown baud rate %s", rate);
    return (sp->speed);
}

 
static void
add_mapping(const char *port, char *arg)
{
    MAP *mapp;
    char *copy, *p;
    const char *termp;
    char *base = 0;

    copy = strdup(arg);
    mapp = typeMalloc(MAP, 1);
    if (copy == 0 || mapp == 0)
	failed("malloc");

    assert(copy != 0);
    assert(mapp != 0);

    mapp->next = 0;
    if (maplist == 0)
	cur = maplist = mapp;
    else {
	cur->next = mapp;
	cur = mapp;
    }

    mapp->porttype = arg;
    mapp->conditional = 0;

    arg = strpbrk(arg, "><@=!:");

    if (arg == 0) {		 
	mapp->type = mapp->porttype;
	mapp->porttype = 0;
	goto done;
    }

    if (arg == mapp->porttype)	 
	termp = mapp->porttype = 0;
    else
	termp = base = arg;

    for (;; ++arg) {		 
	switch (*arg) {
	case '<':
	    if (mapp->conditional & GT)
		goto badmopt;
	    mapp->conditional |= LT;
	    break;
	case '>':
	    if (mapp->conditional & LT)
		goto badmopt;
	    mapp->conditional |= GT;
	    break;
	case '@':
	case '=':		 
	    mapp->conditional |= EQ;
	    break;
	case '!':
	    mapp->conditional |= NOT;
	    break;
	default:
	    goto next;
	}
    }

  next:
    if (*arg == ':') {
	if (mapp->conditional)
	    goto badmopt;
	++arg;
    } else {			 
	arg = strchr(p = arg, ':');
	if (arg == 0)
	    goto badmopt;
	*arg++ = '\0';
	mapp->speed = tbaudrate(p);
    }

    mapp->type = arg;

     
    if (termp != 0)
	*base = '\0';

     
    if (mapp->conditional & NOT)
	mapp->conditional = ~mapp->conditional & (EQ | GT | LT);

     
  done:
    if (port) {
	if (mapp->porttype) {
	  badmopt:
	    err("illegal -m option format: %s", copy);
	}
	mapp->porttype = port;
    }
    free(copy);
#ifdef MAPDEBUG
    (void) printf("port: %s\n", mapp->porttype ? mapp->porttype : "ANY");
    (void) printf("type: %s\n", mapp->type);
    (void) printf("conditional: ");
    p = "";
    if (mapp->conditional & GT) {
	(void) printf("GT");
	p = "/";
    }
    if (mapp->conditional & EQ) {
	(void) printf("%sEQ", p);
	p = "/";
    }
    if (mapp->conditional & LT)
	(void) printf("%sLT", p);
    (void) printf("\nspeed: %d\n", mapp->speed);
#endif
}

 
static const char *
mapped(const char *type)
{
    MAP *mapp;
    int match;

    for (mapp = maplist; mapp; mapp = mapp->next)
	if (mapp->porttype == 0 || !strcmp(mapp->porttype, type)) {
	    switch (mapp->conditional) {
	    case 0:		 
		match = TRUE;
		break;
	    case EQ:
		match = ((int) ospeed == mapp->speed);
		break;
	    case GE:
		match = ((int) ospeed >= mapp->speed);
		break;
	    case GT:
		match = ((int) ospeed > mapp->speed);
		break;
	    case LE:
		match = ((int) ospeed <= mapp->speed);
		break;
	    case LT:
		match = ((int) ospeed < mapp->speed);
		break;
	    default:
		match = FALSE;
	    }
	    if (match)
		return (mapp->type);
	}
     
    return (type);
}

 

 
static const char *
get_termcap_entry(int fd, char *userarg)
{
    int errret;
    char *p;
    const char *ttype;
#if HAVE_PATH_TTYS
#if HAVE_GETTTYNAM
    struct ttyent *t;
#else
    FILE *fp;
#endif
    char *ttypath;
#endif  

    (void) fd;

    if (userarg) {
	ttype = userarg;
	goto found;
    }

     
    if ((ttype = getenv("TERM")) != 0)
	goto map;

#if HAVE_PATH_TTYS
    if ((ttypath = ttyname(fd)) != 0) {
	p = _nc_basename(ttypath);
#if HAVE_GETTTYNAM
	 
	if ((t = getttynam(p))) {
	    ttype = t->ty_type;
	    goto map;
	}
#else
	if ((fp = fopen("/etc/ttytype", "r")) != 0
	    || (fp = fopen("/etc/ttys", "r")) != 0) {
	    char buffer[BUFSIZ];
	    char *s, *t, *d;

	    while (fgets(buffer, sizeof(buffer) - 1, fp) != 0) {
		for (s = buffer, t = d = 0; *s; s++) {
		    if (isspace(UChar(*s)))
			*s = '\0';
		    else if (t == 0)
			t = s;
		    else if (d == 0 && s != buffer && s[-1] == '\0')
			d = s;
		}
		if (t != 0 && d != 0 && !strcmp(d, p)) {
		    ttype = strdup(t);
		    fclose(fp);
		    goto map;
		}
	    }
	    fclose(fp);
	}
#endif  
    }
#endif  

     
    ttype = "unknown";

  map:ttype = mapped(ttype);

     
  found:
    if ((p = getenv("TERMCAP")) != 0 && !_nc_is_abs_path(p)) {
	 
	int n;
	for (n = 0; environ[n] != 0; n++) {
	    if (!strncmp("TERMCAP=", environ[n], (size_t) 8)) {
		while ((environ[n] = environ[n + 1]) != 0) {
		    n++;
		}
		break;
	    }
	}
    }

     
    if (ttype[0] == '?') {
	if (ttype[1] != '\0')
	    ttype = askuser(ttype + 1);
	else
	    ttype = askuser(0);
    }
     
    while (setupterm((NCURSES_CONST char *) ttype, fd, &errret)
	   != OK) {
	if (errret == 0) {
	    (void) fprintf(stderr, "%s: unknown terminal type %s\n",
			   _nc_progname, ttype);
	    ttype = 0;
	} else {
	    (void) fprintf(stderr,
			   "%s: can't initialize terminal type %s (error %d)\n",
			   _nc_progname, ttype, errret);
	    ttype = 0;
	}
	ttype = askuser(ttype);
    }
#if BROKEN_LINKER
    tgetflag("am");		 
#endif
    return (ttype);
}

 

 
static void
obsolete(char **argv)
{
    for (; *argv; ++argv) {
	char *parm = argv[0];

	if (parm[0] == '-' && parm[1] == '\0') {
	    argv[0] = strdup("-q");
	    continue;
	}

	if ((parm[0] != '-')
	    || (argv[1] && argv[1][0] != '-')
	    || (parm[1] != 'e' && parm[1] != 'i' && parm[1] != 'k')
	    || (parm[2] != '\0'))
	    continue;
	switch (argv[0][1]) {
	case 'e':
	    argv[0] = strdup("-e^H");
	    break;
	case 'i':
	    argv[0] = strdup("-i^C");
	    break;
	case 'k':
	    argv[0] = strdup("-k^U");
	    break;
	}
    }
}

static void
print_shell_commands(const char *ttype)
{
    const char *p;
    int len;
    char *var;
    char *leaf;
     
    if ((var = getenv("SHELL")) != 0
	&& ((len = (int) strlen(leaf = _nc_basename(var))) >= 3)
	&& !strcmp(leaf + len - 3, "csh"))
	p = "set noglob;\nsetenv TERM %s;\nunset noglob;\n";
    else
	p = "TERM=%s;\n";
    (void) printf(p, ttype);
}

static void
usage(void)
{
#define SKIP(s)			 
#define KEEP(s) s "\n"
    static const char msg[] =
    {
	KEEP("")
	KEEP("Options:")
	SKIP("  -a arpanet  (obsolete)")
	KEEP("  -c          set control characters")
	SKIP("  -d dialup   (obsolete)")
	KEEP("  -e ch       erase character")
	KEEP("  -I          no initialization strings")
	KEEP("  -i ch       interrupt character")
	KEEP("  -k ch       kill character")
	KEEP("  -m mapping  map identifier to type")
	SKIP("  -p plugboard (obsolete)")
	KEEP("  -Q          do not output control key settings")
	KEEP("  -q          display term only, do no changes")
	KEEP("  -r          display term on stderr")
	SKIP("  -S          (obsolete)")
	KEEP("  -s          output TERM set command")
	KEEP("  -V          print curses-version")
	KEEP("  -w          set window-size")
	KEEP("")
	KEEP("If neither -c/-w are given, both are assumed.")
    };
#undef KEEP
#undef SKIP
    (void) fprintf(stderr, "Usage: %s [options] [terminal]\n", _nc_progname);
    fputs(msg, stderr);
    ExitProgram(EXIT_FAILURE);
     
}

static char
arg_to_char(void)
{
    return (char) ((optarg[0] == '^' && optarg[1] != '\0')
		   ? ((optarg[1] == '?') ? '\177' : CTRL(optarg[1]))
		   : optarg[0]);
}

int
main(int argc, char **argv)
{
    int ch, noinit, noset, quiet, Sflag, sflag, showterm;
    const char *ttype;
    int terasechar = -1;	 
    int intrchar = -1;		 
    int tkillchar = -1;		 
    int my_fd;
    bool opt_c = FALSE;		 
    bool opt_w = FALSE;		 
    TTY mode, oldmode;

    _nc_progname = _nc_rootname(*argv);
    obsolete(argv);
    noinit = noset = quiet = Sflag = sflag = showterm = 0;
    while ((ch = getopt(argc, argv, "a:cd:e:Ii:k:m:p:qQrSsVw")) != -1) {
	switch (ch) {
	case 'c':		 
	    opt_c = TRUE;
	    break;
	case 'a':		 
	    add_mapping("arpanet", optarg);
	    break;
	case 'd':		 
	    add_mapping("dialup", optarg);
	    break;
	case 'e':		 
	    terasechar = arg_to_char();
	    break;
	case 'I':		 
	    noinit = 1;
	    break;
	case 'i':		 
	    intrchar = arg_to_char();
	    break;
	case 'k':		 
	    tkillchar = arg_to_char();
	    break;
	case 'm':		 
	    add_mapping(0, optarg);
	    break;
	case 'p':		 
	    add_mapping("plugboard", optarg);
	    break;
	case 'Q':		 
	    quiet = 1;
	    break;
	case 'q':		 
	    noset = 1;
	    break;
	case 'r':		 
	    showterm = 1;
	    break;
	case 'S':		 
	    Sflag = 1;
	    break;
	case 's':		 
	    sflag = 1;
	    break;
	case 'V':		 
	    puts(curses_version());
	    ExitProgram(EXIT_SUCCESS);
	case 'w':		 
	    opt_w = TRUE;
	    break;
	case '?':
	default:
	    usage();
	}
    }

    argc -= optind;
    argv += optind;

    if (argc > 1)
	usage();

    if (!opt_c && !opt_w)
	opt_c = opt_w = TRUE;

    my_fd = save_tty_settings(&mode, TRUE);
    oldmode = mode;
#ifdef TERMIOS
    ospeed = (NCURSES_OSPEED) cfgetospeed(&mode);
#elif defined(EXP_WIN32_DRIVER)
    ospeed = 0;
#else
    ospeed = (NCURSES_OSPEED) mode.sg_ospeed;
#endif

    if (same_program(_nc_progname, PROG_RESET)) {
	reset_start(stderr, TRUE, FALSE);
	reset_tty_settings(my_fd, &mode, noset);
    } else {
	reset_start(stderr, FALSE, TRUE);
    }

    ttype = get_termcap_entry(my_fd, *argv);

    if (!noset) {
#if HAVE_SIZECHANGE
	if (opt_w) {
	    set_window_size(my_fd, &lines, &columns);
	}
#endif
	if (opt_c) {
	    set_control_chars(&mode, terasechar, intrchar, tkillchar);
	    set_conversions(&mode);

	    if (!noinit) {
		if (send_init_strings(my_fd, &oldmode)) {
		    (void) putc('\r', stderr);
		    (void) fflush(stderr);
		    (void) napms(1000);		 
		}
	    }

	    update_tty_settings(&oldmode, &mode);
	}
    }

    if (noset) {
	(void) printf("%s\n", ttype);
    } else {
	if (showterm)
	    (void) fprintf(stderr, "Terminal type is %s.\n", ttype);
	 
	if (!quiet) {
	    print_tty_chars(&oldmode, &mode);
	}
    }

    if (Sflag)
	err("The -S option is not supported under terminfo.");

    if (sflag) {
	print_shell_commands(ttype);
    }

    ExitProgram(EXIT_SUCCESS);
}
